backends/nats/nats.go

```diff
@@ -0,0 +1,486 @@
+	DisableRetryOnFailedConnect bool `yaml:"disableRetryOnFailedConnect"`
+	MaxReconnects               int  `yaml:"maxReconnects"`
+	ReconnectWaitSeconds        int  `yaml:"reconnectWait"`
+	MaxWaitSeconds              int  `yaml:"maxWaitSeconds"`
+	// Authentication options
+	NatsUsername            string `yaml:"natsUsername"`
+	NatsPassword            string `yaml:"natsPassword"`
+	NatsToken               string `yaml:"natsToken"`
+	NatsNkeySeedFilePath    string `yaml:"natsNkeySeedFilePath"`
+	NatsCredentialsFilePath string `yaml:"natsCredentialsFilePath"`
+	NatsJWTKeyFilePath      string `yaml:"natsJWTKeyFilePath"`
+	// TLS Options
+	// NatsTLSRootCA Specify a custom root CA file (e.g. for self-signed certs)
+	NatsTLSRootCA string `yaml:"natsTLSRootCA"`
+	// NatsTLSClientCert Specify a client cert file
+	NatsTLSClientCert string `yaml:"natsTLSClientCert"`
+	// NatsTLSClientKey Specify a client key file
+	NatsTLSClientKey string `yaml:"natsTLSClientKey"`
+	// Connection options
+	NatsURL string `yaml:"natsURL"`
+	// Storage Options
+	// NatsBucket defines the bucket name
+	NatsBucket string `yaml:"natsBucket"`
+	// NatsBucketReplicas defines number of replicas
+	NatsBucketReplicas int `yaml:"natsBucketReplicas"`
+	// NatsBucketDescription description for NatsBucket
+	NatsBucketDescription string `yaml:"natsBucketDescription"`
+	// CreateBucket tells us to try to create the bucket
+	CreateBucket bool `yaml:"create_bucket"`
+	// CreateBucketPlacementCluster optionally set the cluster placement value when creating a bucket
+	CreateBucketPlacementCluster string `yaml:"createBucketPlacementCluster"`
+	// CreateBucketPlacementTagList comma seperated list of optional placement tags when creating a bucket
+	CreateBucketPlacementTagList string `yaml:"createBucketTagList"`
+	// GlobalPrefix is a prefix applied to all operations, allowing work within a prefix
+	GlobalPrefix string `yaml:"global_prefix"`
+	// EncryptionKey -  key (hex format,256 bit) for encryption at rest
+	EncryptionKey string `yaml:"encryptionKey"`
+	// S3 Compatability
+	// PrefixFolders can be enabled to make List operations show nested prefixes as folders
+	// instead of recursively listing all contents of nested prefixes
+	PrefixFolders bool `yaml:"prefix_folders"`
+	// Internally managed options
+	// Auth type flag set internally by checkCredentialsAvailability()
+	internalUseAuthType authType
+	// Converted type for wait seconds
+	internalReconnectWaitSeconds time.Duration
+	// Converted type for max wait
+	internalMaxWaitSeconds time.Duration
+	// Converted key bytes
+	internalEncryptionKeyBytes []byte
+	// Not loaded from YAML
+	Logger logr.Logger `yaml:"-"`
+}
+
+type Backend struct {
+	opt        Options
+	nc         *nats.Conn
+	log        logr.Logger
+	markerName string
+}
+
+// Check TLS params are available
+func (o Options) checkTLS() error {
+	if o.NatsTLSRootCA != "" {
+		err := fileExistsAndIsReadable(o.NatsTLSRootCA)
+		if err != nil {
+			return err
```

Comment:
You may want to wrap this in a hint about which option is incorrect. Same elsewhere.

Reply:
Just FYI, my intention with this and most of the NATS options was "you know if you need them", i.e. you know if your environment is using an in-house CA (and therefore you need to provide the root pubkey), you know what auth-type your environment is using (and therefore you know if you need to provide an NKey or a username/password) etc.